doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
            intros; absurd (Fail = Fail); trivial.
          Defined.
 
-   .. exn:: Invalid argument.
-
-      The tactic :tacn:`refine` does not know what to do with the term you gave.
-
-   .. exn:: Refine passed ill-formed term.
-
-      The term you gave is not a valid proof (not easy to debug in general). This
-      message may also occur in higher-level tactics that call :tacn:`refine`
-      internally.
-
    .. exn:: Cannot infer a term for this placeholder.
       :name: Cannot infer a term for this placeholder. (refine)
 
       There is a hole in the term you gave whose type cannot be inferred. Put a
       cast around it.
 
-   .. tacv:: simple refine @term
-      :name: simple refine
+   Setting :opt:`Debug` ``"unification"`` enables printing traces of
+   unification steps used during elaboration/typechecking and the
+   :tacn:`refine` tactic. ``"ho-unification"`` prints information
+   about higher order heuristics.
 
-      This tactic behaves like refine, but it does not shelve any subgoal. It does
-      not perform any beta-reduction either.
+.. tacn:: apply {+, {? > } @one_term {? with @bindings } } {? @in_hyp_as }
 
-   .. tacv:: notypeclasses refine @term
-      :name: notypeclasses refine
+   .. insertprodn in_hyp_as as_ipat
 
-      This tactic behaves like :tacn:`refine` except it performs type checking without
-      resolution of typeclasses.
+   .. prodn::
+      in_hyp_as ::= in {+, @ident {? @as_ipat } }
+      as_ipat ::= as @simple_intropattern
+
+   Uses unification to match the type of each :n:`@one_term` with the goal
+   (to do backward reasoning) or with a hypothesis (to do forward reasoning).
+   Specifying multiple :n:`@one_term {? with @bindings }` is equivalent to
+   giving each one serially, left to right, as separate `apply` tactics.
+
+   The type of :n:`@one_term` consists of zero or more :term:`premises <premise>`
+   followed by a :ref:`conclusion <conclusion_meaning_2>`,
+   i.e. it has the form :n:`{* @term__premise -> } @term__conclusion`.
```

Comment:
Can you give an example with a forall and suggest the correct form to show here?  (If it doesn't lend itself to being written out it grammatical form we can modify the text.

Reply:
Basically, a correct way of generalizing this would be to use the dependent arrow instead of the non-dependent arrow, i.e., `{* forall {| @binder | _ : @term_premise }, } @term_conclusion`, where any case of a `forall _ : A` could be rewritten `A ->` by using the non-dependent arrow notation. However, this may not be the most readable form to present to the reader.

This is also not so interesting since statements may be equivalently restated with all `forall`s at the beginning, so instead, we could show the form `forall @binders, {* @term_premise -> } @term_conclusion` and then mention in the text that interleaving of binders and premises is also supported.

Any theorem from the standard library (e.g., `Nat.add_comm`) that we `apply` is an example.